File: app/src/main/java/org/tvheadend/tvhclient/ui/features/settings/SettingsActivity.java
package org.tvheadend.tvhclient.ui.features.settings;

import android.app.Fragment;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;

import com.afollestad.materialdialogs.folderselector.FolderChooserDialog;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.ui.base.BaseActivity;
import org.tvheadend.tvhclient.ui.base.callbacks.BackPressedInterface;
import org.tvheadend.tvhclient.ui.base.utils.SnackbarMessageReceiver;
import org.tvheadend.tvhclient.ui.features.changelog.ChangeLogActivity;
import org.tvheadend.tvhclient.util.MiscUtils;

import java.io.File;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import timber.log.Timber;

public class SettingsActivity extends BaseActivity implements FolderChooserDialog.FolderCallback {

    private SnackbarMessageReceiver snackbarMessageReceiver;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        setTheme(MiscUtils.getThemeId(this));
        super.onCreate(savedInstanceState);
        setContentView(R.layout.misc_content_activity);
        MiscUtils.setLanguage(this);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        snackbarMessageReceiver = new SnackbarMessageReceiver(this);

        String settingType = "default";
        if (getIntent().hasExtra("setting_type")) {
            settingType = getIntent().getStringExtra("setting_type");
        }

        Fragment fragment;
        if (savedInstanceState == null) {
            Timber.d("Saved instance is null, creating settings fragment");
            fragment = getSettingsFragment(settingType);
            fragment.setArguments(getIntent().getExtras());
        } else {
            Timber.d("Saved instance is not null, trying to find settings fragment");
            fragment = getFragmentManager().findFragmentById(R.id.main);
        }

        Timber.d("Replacing fragment");
        getFragmentManager().beginTransaction().replace(R.id.main, fragment).commit();
    }

    private Fragment getSettingsFragment(@NonNull String type) {
        Timber.d("Getting settings fragment for type '" + type + "'");
        Fragment fragment = null;
        Intent intent;
        switch (type) {
            case "list_connections":
                fragment = new SettingsListConnectionsFragment();
                break;
            case "add_connection":
                fragment = new SettingsAddConnectionFragment();
                break;
            case "edit_connection":
                fragment = new SettingsEditConnectionFragment();
                break;
            case "user_interface":
                fragment = new SettingsUserInterfaceFragment();
                break;
            case "profiles":
                fragment = new SettingsProfilesFragment();
                break;
            case "playback":
                fragment = new SettingsPlaybackFragment();
                break;
            case "advanced":
                fragment = new SettingsAdvancedFragment();
                break;
            case "changelog":
                intent = new Intent(this, ChangeLogActivity.class);
                startActivity(intent);
                break;
            default:
                fragment = new SettingsFragment();
                fragment.setArguments(getIntent().getExtras());
                break;
        }
        return fragment;
    }

    @Override
    public void onStart() {
        super.onStart();
        LocalBroadcastManager.getInstance(this).registerReceiver(snackbarMessageReceiver, new IntentFilter(SnackbarMessageReceiver.ACTION));
    }

    @Override
    public void onStop() {
        super.onStop();
        LocalBroadcastManager.getInstance(this).unregisterReceiver(snackbarMessageReceiver);
    }

    @Override
    public void onBackPressed() {
        // If a settings fragment is currently visible, let the fragment
        // handle the back press, otherwise the setting activity.
        Fragment fragment = getFragmentManager().findFragmentById(R.id.main);
        if (fragment instanceof BackPressedInterface
                && fragment.isVisible()) {
            Timber.d("Calling back press in the fragment");
            ((BackPressedInterface) fragment).onBackPressed();
        } else {
            Timber.d("Calling back press of super");
            super.onBackPressed();
        }
    }

    @Override
    public void onFolderSelection(@NonNull FolderChooserDialog dialog, @NonNull File folder) {
        Fragment f = getFragmentManager().findFragmentById(R.id.main);
        if (f != null && f.isAdded() && f instanceof FolderChooserDialogCallback) {
            ((FolderChooserDialogCallback) f).onFolderSelected(folder);
        }
    }

    @Override
    public void onFolderChooserDismissed(@NonNull FolderChooserDialog dialog) {
        // NOP
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/settings/SettingsActivity.kt
package org.tvheadend.tvhclient.ui.features.settings


import android.content.Context
import android.content.IntentFilter
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import org.tvheadend.tvhclient.BuildConfig
import org.tvheadend.tvhclient.MainApplication
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.ui.common.SnackbarMessageReceiver
import org.tvheadend.tvhclient.ui.common.interfaces.BackPressedInterface
import org.tvheadend.tvhclient.ui.common.interfaces.ToolbarInterface
import org.tvheadend.tvhclient.ui.common.onAttach
import org.tvheadend.tvhclient.ui.features.information.ChangeLogFragment
import org.tvheadend.tvhclient.ui.features.information.InformationFragment
import org.tvheadend.tvhclient.ui.features.information.PrivacyPolicyFragment
import org.tvheadend.tvhclient.ui.features.unlocker.UnlockerFragment
import org.tvheadend.tvhclient.util.extensions.showSnackbarMessage
import org.tvheadend.tvhclient.util.getThemeId
import timber.log.Timber

class SettingsActivity : AppCompatActivity(), RemoveFragmentFromBackstackInterface, ToolbarInterface {

    private lateinit var settingsViewModel: SettingsViewModel
    private lateinit var snackbarMessageReceiver: SnackbarMessageReceiver

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(getThemeId(this))
        super.onCreate(savedInstanceState)
        setContentView(R.layout.misc_content_activity)

        MainApplication.component.inject(this)

        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        settingsViewModel = ViewModelProvider(this)[SettingsViewModel::class.java]
        snackbarMessageReceiver = SnackbarMessageReceiver(settingsViewModel)

        // If the user wants to go directly to a sub setting screen like the connections
        // and not the main settings screen the setting type can be passed here
        if (savedInstanceState == null && intent.hasExtra("setting_type")) {
            val id = intent.getStringExtra("setting_type") ?: "default"
            settingsViewModel.setNavigationMenuId(id)
        }

        settingsViewModel.getNavigationMenuId().observe(this) { event ->
            event.getContentIfNotHandled()?.let {
                Timber.d("New preference selected with id $it, replacing settings fragment")
                supportFragmentManager.beginTransaction()
                    .replace(R.id.main, getSettingsFragment(it))
                    .addToBackStack(null)
                    .commit()
            }
        }

        settingsViewModel.currentServerStatusLiveData.observe(this) { serverStatus ->
            Timber.d("Received live data, server status has changed and is ${if (serverStatus != null) "" else "not "}available")
            if (serverStatus != null) {
                settingsViewModel.currentServerStatus = serverStatus
            }
        }

        settingsViewModel.snackbarMessageLiveData.observe(this) { event ->
            event.getContentIfNotHandled()?.let {
                this.showSnackbarMessage(it)
            }
        }

        settingsViewModel.isUnlockedLiveData.observe(this) { isUnlocked ->
            Timber.d("Received live data, isUnlocked value changed to $isUnlocked")
            settingsViewModel.isUnlocked = isUnlocked || BuildConfig.OVERRIDE_UNLOCKED
        }
    }

    public override fun onStart() {
        super.onStart()
        LocalBroadcastManager.getInstance(this).registerReceiver(snackbarMessageReceiver, IntentFilter(SnackbarMessageReceiver.SNACKBAR_ACTION))
    }

    public override fun onStop() {
        super.onStop()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(snackbarMessageReceiver)
    }

    override fun attachBaseContext(context: Context) {
        super.attachBaseContext(onAttach(context))
    }

    override fun setTitle(title: String) {
        supportActionBar?.title = title
    }

    override fun setSubtitle(subtitle: String) {
        supportActionBar?.subtitle = subtitle
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun getSettingsFragment(type: String): Fragment {
        Timber.d("Getting settings fragment for type '$type'")
        return when (type) {
            "list_connections" -> SettingsListConnectionsFragment()
            "add_connection" -> SettingsAddConnectionFragment()
            "edit_connection" -> SettingsEditConnectionFragment()
            "user_interface" -> SettingsUserInterfaceFragment()
            "profiles" -> SettingsProfilesFragment()
            "playback" -> SettingsPlaybackFragment()
            "advanced" -> SettingsAdvancedFragment()
            "unlocker" -> UnlockerFragment.newInstance(settingsViewModel.isUnlocked)
            "information" -> InformationFragment()
            "privacy_policy" -> PrivacyPolicyFragment()
            "changelog" -> ChangeLogFragment.newInstance(showFullChangelog = true)
            else -> SettingsFragment()
        }
    }

    /*
     * Forwards the back press command to the fragment that implements the to handle the back press.
     * This is usually required when a confirmation dialog shall be shown before exiting the fragment.
     * Otherwise remove the last fragment from the back stack or close the activity in case only one fragment remains.
     */
    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        val fragment = supportFragmentManager.findFragmentById(R.id.main)
        if (fragment is BackPressedInterface && fragment.isVisible) {
            Timber.d("Calling back press in the fragment")
            fragment.onBackPressed()
        } else {
            Timber.d("Calling back press of super")
            removeFragmentFromBackstack()
        }
    }

    override fun removeFragmentFromBackstack() {
        Timber.d("Back stack count is ${supportFragmentManager.backStackEntryCount}")
        if (supportFragmentManager.backStackEntryCount <= 1) {
            finish()
        } else {
            super.onBackPressed()
        }
    }
}
